File: 2023/lib/dir.ml
type t = Up | Left | Right | Down

let neg = function Up -> Down | Left -> Right | Right -> Left | Down -> Up

let of_char = function
  | '^' -> Up
  | '<' -> Left
  | '>' -> Right
  | 'v' -> Down
  | _ -> invalid_arg __FUNCTION__

let to_char = function Up -> '^' | Left -> '<' | Right -> '>' | Down -> 'v'
let to_string = function Up -> "↑" | Left -> "←" | Down -> "↓" | Right -> "→"
let turn_left = function Up -> Left | Left -> Down | Down -> Right | Right -> Up
let turn_right = function Up -> Right | Left -> Up | Down -> Left | Right -> Down

File: 2021/lib/dir.go
package lib

type Dir int

const (
	Up Dir = iota
	Down
	Left
	Right
)

var AllDir = []Dir{Up, Down, Left, Right}

func (d Dir) Neg() Dir {
	return [4]Dir{
		Up:    Down,
		Down:  Up,
		Left:  Right,
		Right: Left,
	}[d]
}

func (d Dir) ToVector() Vector2 {
	return [4]Vector2{
		Up:    Vector2{Y: -1},
		Down:  Vector2{Y: +1},
		Left:  Vector2{X: -1},
		Right: Vector2{X: +1},
	}[d]
}
